### PR TITLE
feat: mention struct update syntax

### PR DESCRIPTION
**Pull request by fitzchivalrik**
When talking about the `with` expression in C#,
mention the struct update syntax[^1] as something
similar in Rust.

[^1]: https://doc.rust-lang.org/stable/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax

### CODE REVIEW

**Review by atifaziz — CHANGES_REQUESTED** [review]
Thanks. The struct update syntax is something certainly worth mentioning, but I don't think it's comparable to `with`. For example, if `Point` is part of some module then the fields `x` and `y` would be private to the code that's external to the module. The `Point ` would then have to be defined as:

```rust
mod points {
    #[derive(Debug)]
    pub struct Point { pub x: i32, pub y: i32 }
}

fn main() {
    use points::Point;
    let pt = Point { x: 123, y: 456 };
    let pt = Point { x: 789, ..pt };
    println!("{pt:?}"); // prints: Point { x: 789, y: 456 }
}
```

to use the struct update syntax from outside the module. Also note that if you make the fields public then the whole `impl` block can be skipped. Consequently, the example of emulation is still valid so I would leave it intact, but then your bits about the struct update syntax with the above example. I would then point out that it's generally more common to see that use within the module that has access to private details of its types.

**Comment by fitzchivalrik** [comment]
> For example, if Point if part of some module then the fields x and y would be private to the code that's external to the module. 

Yeah, I was debating whether that should be mentioned explicitly, or if it would complicate things too much.
I decided to not mention it, because I thought it is the same for `with`; as in, the fields need to be accessible at call site to be able to use `with`:

```
using System;
					
public readonly record struct Point {
	internal readonly int X {get; init;}
	private readonly int Y {get; init;}
	public Point(int x, int y) => (X,Y)=(x,y);
	public override String ToString() => $"Point {{ X = {this.X}, Y= {this.Y} }}";
}
public class Program
{
	public static void Main()
	{
		var pt = new Point(123, 456);
		pt = pt with { X = 789 }; // Error if `X` is declared private above, or if `Main()` is in another project/module.
		Console.WriteLine(pt.ToString()); // prints: Point { X = 789, Y = 456 }
		
	}
}
```

**Comment by atifaziz** [comment]
> I thought it is the same for `with`; as in, the fields need to be accessible at call site to be able to use `with`

Right, but in the C# example, `X` and `Y` are _properties_, not _fields_, and so it's comparing apples to oranges.

**Comment by fitzchivalrik** [comment]
>  but in the C# example, X and Y are _properties_, not _fields_, 

Ah, yeah, fair point! Thank you!

File: src/language/custom-types/members.md [hunk 1]
```diff
@@ -329,3 +329,22 @@ impl Point {
     pub fn set_y(self, val: i32) -> Self { Self::new(self.x, val) }
 }
 ```
+
+It is also possible to use the struct update syntax to achieve something similar. This moves the original value:
```
Comment (atifaziz) [inline src/language/custom-types/members.md:333]: I'd suggest to reword this to avoid saying it can achieve something similar. I'd instead say it _appears_ to offer something similar, but is actually different and how:

```suggestion
Rust also has a _[struct update syntax]_ that may seem similar to `with` in
C#, but it has a very different behaviour, works fields only and _moves_ the
original value:
```

File: src/language/custom-types/members.md [hunk 2]
```diff
@@ -329,3 +329,22 @@ impl Point {
...
+fn main() {
+    use points::Point;
+    let pt = Point { x: 123, y: 456 };
+    let pt = Point { x: 789, ..pt };
+    println!("{pt:?}"); // prints: Point { x: 789, y: 456 }
+}
+```
+
+As seen in the example above, the struct update syntax requires access to the type fields.
+Therefore, it is generally more common to use it within the module that has access to private details of its types.
```
Comment (atifaziz) [inline src/language/custom-types/members.md:350]: Wrapping and adding a link to the [struct update syntax](https://doc.rust-lang.org/stable/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax) bit in the Rust book:

```suggestion
As seen in the example above, the struct update syntax requires access to the
type's fields. Therefore, it is generally more common to see it used within
the module that has access to private details of its types.

  [struct update syntax]: https://doc.rust-lang.org/stable/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax
```

**Review by atifaziz — CHANGES_REQUESTED** [review]
I've made a few suggestions then we're good to merge. 🏁

**Review by atifaziz — COMMENTED** [review]
I just had a thought. Perhaps it's best to compare the struct update syntax of Rust with [non-destructive mutation](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct#nondestructive-mutation) of _non-record_ `struct` types using `with`, which has been possible since C# 10. It seems a lot more comparable. The difference is that C# does a non-destructive mutation (copy then update) whereas Rust does (partial) moves. I would suggest starting with a C# example of `Point` as a `struct` with public and (read-write) fields:

```c#
var pt = new Point { X = 123, Y = 456 };
Console.WriteLine(pt.ToString()); // prints: (123, 456)
pt = pt with { X = 789 };
Console.WriteLine(pt.ToString()); // prints: (789, 456)

struct Point
{
    public int X;
    public int Y;

    public override string ToString() => $"({X}, {Y})";
}
```

Then go on to show the same is possible in Rust using the struct update syntax and end on the differences.

**Review by atifaziz — APPROVED** [review]
Sorry this took so long, @fitzchivalrik, but I integrated the final edits with d655d1bbc2a9149d7561ced78fc3847a0b8b562b and this is good to go now!

Thanks for contribution and working through the review!

**Comment by atifaziz** [comment]
@fitzchivalrik If you could [agree to the CLA](https://github.com/microsoft/rust-for-dotnet-devs/pull/37#issuecomment-1500910155) then I can get this merged. Thanks!

**Comment by fitzchivalrik** [comment]
@atifaziz No worries, thanks for coming back to it.
Unfortunately, I do not really want to agree to the CLA.
(I am unsure if this contribution even passes the threshold of originality, but I am not a lawyer, of course.)

If this causes complications, I would be open to agree to the CLA _only for this specific contribution_ (and not in general), if this is at all possible.

**Comment by atifaziz** [comment]
> If this causes complications, I would be open to agree to the CLA _only for this specific contribution_ (and not in general), if this is at all possible.

@fitzchivalrik I understand. How about this? You sign it now for this contribution and if you feel uncomfortable with the CLA at any point thereafter and want to revoke it for future contributions, you always have the option to [terminate the agreement](https://github.com/microsoft/contributorlicenseagreement?tab=readme-ov-file#terminating) by issuing the following command:

    @microsoft-github-policy-service terminate

However, I'd encourage you to [seek clarity on the CLA](https://github.com/microsoft/ContributorLicenseAgreement) so you're eventually comfortable with it because we appreciate and value your contributions!

**Comment by fitzchivalrik** [comment]
Sorry for the delay. Thanks for the terminate command - agreeing, and then terminating it later here is fine for me.
@microsoft-github-policy-service agree

**Comment by fitzchivalrik** [comment]
@microsoft-github-policy-service terminate
